### DISCUSSION TITLE
Focus switching with keyboard?

### DISCUSSION THREAD

**Discussion by BillScherer** [Textualize/textual#107] [Q&A] [answered]
Sorry if I missed this somewhere, but is there a way to enable using the Tab key, alone or in combination with a modifier, to cycle focus from one widget to the next?

I'm trying to get something working in a terminal with no mouse support, and there doesn't seem to be any easy way to enable this.

Thanks!

**Comment by willmcgugan** [comment, marked as answer]
There's no tab focus at the moment, but it is an important feature, because I would like Textual to be keyboard first.

What I was thinking is to have an automatic widget order left to right, top to buttom, while also allowing a tab_order attribute to sort by.

There are a few details nail down before I get there. For instance there needs to be someway for a widget to indicate it has focus. I was thinking of a overlayed border that appears while you are tabbing.

**Reply by kfmfe04** [reply]
On a related note (traversing a table with keystrokes):

When I used `on_key()` to capture a "down" key, I attempted to set `table.row_style` directly (set the cursor/active row to "reversed"), but this didn't trigger any refresh/redraw to my table.

For this kind of functionality, should I be using Live to trigger a refresh?
What is the best approach?  Thanks in advance.

Edit:  I found the sample code in rich/live.py - it looks like I should use live and just redraw the entire table, as needed.  I'll experiment with this over the next few days.

**Comment by lllama** [comment]
Pretty sure this doesn't apply here but in web-land, `tabindex` is discouraged for accessibility concerns: https://www.tpgi.com/using-the-tabindex-attribute/

I can't think how this would affect TUI apps but then I also know nothing about how screen-readers etc interact with the console. Just an FYI more than anything.
